feat: [CAT-85] Support RMV I/O descriptions in CLI

marcelo-verta-ai opened the pull request:
# Summary

Modify python client to accept new attributes

# References

https://vertaai.atlassian.net/browse/CAT-85?atlOrigin=eyJpIjoiNjk0YTRlYWU2MDgzNDVkMGE3YjVkZDI5N2ZiNGVmODIiLCJwIjoiaiJ9
# Dependencies

https://vertaai.atlassian.net/browse/CAT-89?atlOrigin=eyJpIjoiMGY5NmQ3ZjNiMWEzNGQyNzhiYmNiODZmNzM3YjkyMzgiLCJwIjoiaiJ9

# Test Result (2022-09-19 16:48)
```
export VERTA_HOST="https://modelcatalog.dev.verta.ai/"
export VERTA_EMAIL="marcelo@verta.ai"
export VERTA_DEV_KEY="7ae98d72-d0dd-4034-9fb1-e2f79e894fd9"
pytest -k "TestModelIODescription" -vvra
pytest -k "TestModelIODescription" -vvra
================================================================================== test session starts ===================================================================================
platform linux -- Python 3.9.14, pytest-7.1.3, pluggy-1.0.0 -- /usr/bin/python3.9
cachedir: .pytest_cache
hypothesis profile 'default' -> suppress_health_check=[HealthCheck.too_slow], database=DirectoryBasedExampleDatabase('/opt/repositorio-git/verta/modeldb/client/verta/tests/.hypothesis/examples')
rootdir: /opt/repositorio-git/verta/modeldb/client/verta/tests, configfile: pytest.ini
plugins: forked-1.4.0, hypothesis-6.54.6, xdist-2.5.0
collected 1309 items / 1304 deselected / 2 skipped / 5 selected                                                                                                                          

registry/model_version/test_crud.py::TestModelIODescription::test_input_description PASSED                                                                                         [ 20%]
registry/model_version/test_crud.py::TestModelIODescription::test_output_description PASSED                                                                                        [ 40%]
registry/model_version/test_crud.py::TestModelIODescription::test_hide_input_label PASSED                                                                                          [ 60%]
registry/model_version/test_crud.py::TestModelIODescription::test_hide_output_label PASSED                                                                                         [ 80%]
test_cli/test_registry.py::TestModelIODescription::test_create_version PASSED                                                                                                      [100%]

==================================================================================== warnings summary ====================================================================================
../verta/registry/data_type/__init__.py:4
  /opt/repositorio-git/verta/modeldb/client/verta/verta/registry/data_type/__init__.py:4: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

test_utils/test_pip_requirements.py:156
  /opt/repositorio-git/verta/modeldb/client/verta/tests/test_utils/test_pip_requirements.py:156: NonInteractiveExampleWarning: The `.example()` method is good for exploring strategies, but should only be used interactively.  We recommend using `@given` for tests - it performs better, saves and replays failures to avoid flakiness, and reports minimal examples. (strategy: versions())
    version=versions().example(),  # pylint: disable=no-member

registry/model_version/test_crud.py::TestModelIODescription::test_input_description
registry/model_version/test_crud.py::TestModelIODescription::test_input_description
registry/model_version/test_crud.py::TestModelIODescription::test_output_description
registry/model_version/test_crud.py::TestModelIODescription::test_hide_input_label
registry/model_version/test_crud.py::TestModelIODescription::test_hide_output_label
test_cli/test_registry.py::TestModelIODescription::test_create_version
test_cli/test_registry.py::TestModelIODescription::test_create_version
test_cli/test_registry.py::TestModelIODescription::test_create_version
test_cli/test_registry.py::TestModelIODescription::test_create_version
  /opt/repositorio-git/verta/modeldb/client/verta/verta/_internal_utils/_utils.py:80: DeprecationWarning: Using 'method_whitelist' with Retry is deprecated and will be removed in v2.0. Use 'allowed_methods' instead
    self.retry = Retry(total=max_retries,

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================================================================ short test summary info =================================================================================
SKIPPED [1] test_backend.py:6: reduce DB pressure to establish test baseline
SKIPPED [1] test_datasets.py:3: outdated tests for legacy API
========================================================= 5 passed, 2 skipped, 1304 deselected, 11 warnings in 89.64s (0:01:29) ==========================================================
```

marcelo-verta-ai committed:
feat: added new fields input_description, output_description, hide_input_label and hide_output_label on model_version

marcelo-verta-ai committed:
Merge branch 'main' into marcelo/cat-85

# Conflicts:
#	client/verta/verta/_protos/public/registry/RegistryService_pb2.py

liuverta commented on client/verta/verta/_cli/registry/create.py:
```diff
@@ -68,10 +68,15 @@ def create_model(model_name, label, visibility, workspace, description, task_typ
+@click.option("--input_description", "-d", help="Input description.")
+@click.option("--hide_input_label", help="Flag to hide input label.", is_flag=True)
+@click.option("--output_description", "-d", help="Output description.")
+@click.option("--hide_output_label", help="Flag to hide output label.", is_flag=True)
```
We cannot re-use the short name `-d` because it is already being used by `--description`. Option names `--<foo>` and `-<bar>` have to be unique for a command.

Also, for CLI, we should use `-` in the option names instead of `_`.

```suggestion
@click.option("--input-description", help="Input description.")
@click.option("--hide-input-label", help="Flag to hide input label.", is_flag=True)
@click.option("--output-description", help="Output description.")
@click.option("--hide-output-label", help="Flag to hide output label.", is_flag=True)
```

liuverta commented on client/verta/verta/registry/entities/_model.py:
```diff
@@ -163,6 +200,14 @@ def create_version(self, name=None, desc=None, labels=None, attrs=None, time_cre
+        output_description : str, optional
+            Output Description of the Model Version.
+        hide_output_label : bool, optional
+            Hide Output Label of the Model Version.
```
```suggestion
        input_description : str, optional
            Description of the model's input.
        hide_input_label : bool, optional
            Whether to hide the model's input label.
        output_description : str, optional
            Description of the model's output.
        hide_output_label : bool, optional
            Whether to hide the model version's output label.
```

liuverta commented on client/verta/verta/registry/entities/_model.py:
```diff
@@ -101,6 +114,14 @@ def get_or_create_version(self, name=None, desc=None, labels=None, attrs=None, t
+        input_description : str, optional
+            Input Description of the Model Version.
+        hide_input_label : bool, optional
+            Hide Input Label of the Model Version.
```
It feels strange to me for `hide_input_label` and `hide_output_label` to be in the client/CLI at all, because showing labels is entirely a UI feature, right? 🤔 

If we must have them in the client, I would recommend

```suggestion
        hide_input_label : bool, default False
            Hide Input Label of the Model Version.
```

instead of `optional` (`=None`) for booleans.

liuverta reviewed (changes requested):
New client tests should go in https://github.com/VertaAI/modeldb/blob/main/client/verta/tests/registry/model_version/test_crud.py. Please make a new `class TestModelIODescription` in that file, and maybe use `test_labels` as a guide. And please also add a check in `test_repr`.

New CLI tests should go in https://github.com/VertaAI/modeldb/blob/main/client/verta/tests/test_cli/test_registry.py.

dependabot[bot] committed:
chore(deps): bump aws.version from 1.12.300 to 1.12.301 (#3215)

Bumps `aws.version` from 1.12.300 to 1.12.301.

Updates `aws-java-sdk-s3` from 1.12.300 to 1.12.301
- [Release notes](https://github.com/aws/aws-sdk-java/releases)
- [Changelog](https://github.com/aws/aws-sdk-java/blob/master/CHANGELOG.md)
- [Commits](https://github.com/aws/aws-sdk-java/compare/1.12.300...1.12.301)

Updates `aws-java-sdk-sts` from 1.12.300 to 1.12.301
- [Release notes](https://github.com/aws/aws-sdk-java/releases)
- [Changelog](https://github.com/aws/aws-sdk-java/blob/master/CHANGELOG.md)
- [Commits](https://github.com/aws/aws-sdk-java/compare/1.12.300...1.12.301)

---
updated-dependencies:
- dependency-name: com.amazonaws:aws-java-sdk-s3
  dependency-type: direct:production
  update-type: version-update:semver-patch
- dependency-name: com.amazonaws:aws-java-sdk-sts
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): bump spotless-maven-plugin from 2.25.0 to 2.26.0 (#3218)

Bumps [spotless-maven-plugin](https://github.com/diffplug/spotless) from 2.25.0 to 2.26.0.
- [Release notes](https://github.com/diffplug/spotless/releases)
- [Changelog](https://github.com/diffplug/spotless/blob/main/CHANGES.md)
- [Commits](https://github.com/diffplug/spotless/compare/lib/2.25.0...lib/2.26.0)

---
updated-dependencies:
- dependency-name: com.diffplug.spotless:spotless-maven-plugin
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): bump aws.version from 1.12.301 to 1.12.302 (#3216)

Bumps `aws.version` from 1.12.301 to 1.12.302.

Updates `aws-java-sdk-s3` from 1.12.301 to 1.12.302
- [Release notes](https://github.com/aws/aws-sdk-java/releases)
- [Changelog](https://github.com/aws/aws-sdk-java/blob/master/CHANGELOG.md)
- [Commits](https://github.com/aws/aws-sdk-java/compare/1.12.301...1.12.302)

Updates `aws-java-sdk-sts` from 1.12.301 to 1.12.302
- [Release notes](https://github.com/aws/aws-sdk-java/releases)
- [Changelog](https://github.com/aws/aws-sdk-java/blob/master/CHANGELOG.md)
- [Commits](https://github.com/aws/aws-sdk-java/compare/1.12.301...1.12.302)

---
updated-dependencies:
- dependency-name: com.amazonaws:aws-java-sdk-s3
  dependency-type: direct:production
  update-type: version-update:semver-patch
- dependency-name: com.amazonaws:aws-java-sdk-sts
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

liuverta committed:
chore!: Remove support for Python 3.6 (#3220)

marcelo-verta-ai committed:
Update client/verta/verta/_cli/registry/create.py

Co-authored-by: Liu <96442646+liuverta@users.noreply.github.com>

marcelo-verta-ai committed:
Update client/verta/verta/registry/entities/_model.py

Co-authored-by: Liu <96442646+liuverta@users.noreply.github.com>

marcelo-verta-ai committed:
fix: requested changes by @liuverta

marcelo-verta-ai committed:
Merge branch 'main' into marcelo/cat-85

marcelo-verta-ai committed:
fix: defined default value for hide_input_label and hide_output_label

marcelo-verta-ai committed:
feat: unit tests for model version input/output

marcelo-verta-ai committed:
fix: removed wrong semicolon

marcelo-verta-ai committed:
feat: unit tests for model version input/output

liuverta commented on client/verta/verta/registry/entities/_model.py:
```diff
@@ -144,10 +169,22 @@ def set_version(self, *args, **kwargs):
         Alias for :meth:`RegisteredModel.get_or_create_version()`.
 
-        """
+        """                                                                                                                                                                                                                                                                                                                                                                                                     ...
```
This line seems to have 1200 extra spaces 😅 

```suggestion
        """
```

liuverta commented on client/verta/verta/_cli/registry/update.py:
```diff
@@ -58,9 +58,14 @@ def update_model(model_name, label, workspace, description):
+@click.option("--input_description", "-d", help="Input description.")
+@click.option("--hide_input_label", help="Flag to hide input label.", is_flag=True)
+@click.option("--output_description", "-d", help="Output description.")
+@click.option("--hide_output_label", help="Flag to hide output label.", is_flag=True)
```
```suggestion
@click.option("--input-description", help="Input description.")
@click.option("--hide-input-label", help="Flag to hide input label.", is_flag=True)
@click.option("--output-description", help="Output description.")
@click.option("--hide-output-label", help="Flag to hide output label.", is_flag=True)
```

liuverta commented on client/verta/verta/_cli/registry/create.py:
```diff
@@ -89,7 +94,15 @@ def create_model_version(ctx, model_name, version_name, label, model, custom_mod
+        input_description=input_description,
+        hide_input_label=hide_input_label,
+        output_description=output_description,
+        hide_output_label=hide_output_label
```
```suggestion
        hide_output_label=hide_output_label,
```

liuverta commented on client/verta/tests/registry/model_version/test_crud.py:
```diff
@@ -378,3 +385,37 @@ def test_batch(self, model_version, strs):
+    def test_input_description(self, client, created_entities):
+        registered_model = client.set_registered_model()
+        created_entities.append(registered_model)
+        model_version = registered_model.get_or_create_version(name="my version")
```
I know this might be a bit confusing and existing tests are a bit inconsistent 😭 but we should try using `create_*()` when creating new entities.

```suggestion
        registered_model = client.create_registered_model()
        created_entities.append(registered_model)
        model_version = registered_model.create_version(name="my version")
```

liuverta commented on client/verta/tests/registry/model_version/test_crud.py:
```diff
@@ -378,3 +385,37 @@ def test_batch(self, model_version, strs):
+        model_version = registered_model.get_or_create_version(name="my version")
+        desc = "input description"
+        model_version.set_input_description(desc)
+        assert desc == model_version.get_input_description()
```
For each of these tests, after this `assert`, could you also add

```python
retrieved_model_version = registered_model.get_version(id=model_version.id)
assert desc == retrieved_model_version.get_input_description()
```

This helps ensure that registry is persisting the data.

marcelo-verta-ai replied:
ok

liuverta commented on client/verta/tests/test_cli/test_registry.py:
```diff
@@ -717,3 +717,53 @@ def test_creation(self, client, created_entities, data_type_cls):
+        runner = CliRunner()
+        result = runner.invoke(
+            cli,
+            ['registry', 'create', 'registeredmodel', model_name, '-l', 'label1', '-l', 'label2'],
```
We can skip these labels since we're not testing them.

```suggestion
            ['registry', 'create', 'registeredmodel', model_name],
```

liuverta reviewed (changes requested):
Some small requests

marcelo-verta-ai committed:
Update client/verta/verta/registry/entities/_model.py

Co-authored-by: Liu <96442646+liuverta@users.noreply.github.com>

marcelo-verta-ai committed:
Update client/verta/verta/_cli/registry/update.py

Co-authored-by: Liu <96442646+liuverta@users.noreply.github.com>

marcelo-verta-ai committed:
Update client/verta/verta/_cli/registry/create.py

Co-authored-by: Liu <96442646+liuverta@users.noreply.github.com>

marcelo-verta-ai committed:
Update client/verta/tests/registry/model_version/test_crud.py

Co-authored-by: Liu <96442646+liuverta@users.noreply.github.com>

marcelo-verta-ai committed:
Update client/verta/tests/test_cli/test_registry.py

Co-authored-by: Liu <96442646+liuverta@users.noreply.github.com>

marcelo-verta-ai committed:
fix: PR revision

marcelo-verta-ai committed:
fix: PR revision

liuverta reviewed (approved):
Looks good. I'll merge soon. Thanks!
